Support Complex Dimension Joins

shangyian opened the pull request:
### Summary

This PR adds support for more complex dimension joins, where the joins may have arbitrary custom SQL. The main changes include:
* Dimension link table schema and input/output models
* API endpoint to define a link between a given node and a dimension node
* Addition to the dimension join logic to use these explicitly defined dimension links, as well as those stored on the columns. We may want to consider joining these together so that the logic goes through a single path.
* Modify `get_dimensions` so that it takes into consideration dimensions from the custom SQL dimension links path

### Test Plan

<!-- How did you test your change? -->

- [x] PR has an associated issue: #853 
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

<!-- Any special instructions around deployment? -->

CircArgs commented on datajunction-server/datajunction_server/api/nodes.py:
```diff
@@ -709,6 +712,133 @@ def link_dimension(  # pylint: disable=too-many-arguments
+    # Check that there is a valid join clause
+    if not join_query.select.from_.relations[0].extensions:  # type: ignore
+        raise DJInvalidInputException(
+            f"Provided SQL `{link_input.join_sql}` does contain JOIN clause",
```
Does *_not_ contain?

CircArgs replied:
or is malformed or something right

shangyian replied:
Ah yes, good catch!

CircArgs commented on datajunction-server/datajunction_server/construction/build.py:
```diff
@@ -270,14 +272,22 @@ def join_tables_for_dimensions(
-                    required_dimension_columns,
-                )
+                if dim_node in dimension_links and dimension_links[dim_node]:
+                    join_query = parse(f"SELECT 1 FROM {dimension_links[dim_node]}")
```
This keeps coming up. It shouldn't be too difficult to have dedicated parsers for parts of queries at least the very high-level nodes. Is it worth it though

shangyian replied:
Hmm, I think it's probably not needed for the time being, since there's a pretty easy workaround.

agorajek commented on datajunction-server/datajunction_server/api/nodes.py:
```diff
@@ -709,6 +712,133 @@ def link_dimension(  # pylint: disable=too-many-arguments
+    if existing_link:
+        # Update the existing dimension link
+        is_update = True
+        dimension_link = existing_link[0]
```
I think this `...[0]` only makes sense if we assume (and validate somewhere) that a node can only be joined to a dimension once. But in reality this is not the case... 

What do you think about adding the "role" attribute to the DimensionLink object and using that the user indicator which join link we are talking about? 

I think it is ok to leave the "role" field as an optional string since most dim links would not need it but during validating we should check if a node has unique dim links as denoted by: dim_node_id + role. Then when we convert all the dimension links to this patter we would get our dim roles "for free".

shangyian replied:
This is a great point -- I had forgotten about the dimension roles issue that we discussed previously, and we can incorporate that nicely here. I will update to add the `role` field here.

agorajek commented on datajunction-server/datajunction_server/database/dimensionlink.py:
```diff
@@ -0,0 +1,61 @@
+
+    # Metadata about the join
+    join_type: Mapped[Optional[JoinType]]
+    join_kind: Mapped[JoinKind] = mapped_column(
```
`join_cardinality` is better.

agorajek commented on datajunction-server/datajunction_server/database/dimensionlink.py:
```diff
@@ -0,0 +1,61 @@
+from datajunction_server.models.dimensionlink import JoinKind, JoinType
+
+
+class DimensionLink(Base):  # pylint: disable=too-few-public-methods
```
As I mentioned earlier `role` may be nice to add here.

agorajek reviewed (commented):
@shangyian I think I have only one major thing to consider before merging this in. Namely the "role" concept - please see an in-line comment for that. The rest are nits.

shangyian commented:
@agorajek I added `role` and changed to `join_cardinality`. There's probably more interesting things we could do with dimension roles, but I'll save that for subsequent PRs.

agorajek reviewed (approved):
Looks good!
